cabal-install/src/Distribution/Client/CmdConfigure.hs

```diff
@@ -140,9 +149,101 @@ configureAction flags@NixStyleFlags {..} _extraArgs globalFlags = do
+
+     ++ "It is never necessary to use the 'v2-reconfigure' command. It is "
+     ++ "merely a convenience in cases where you do not want to specify flags "
+     ++ "to 'v2-build' (and other commands) every time and yet do not want "
+     ++ "to alter the 'cabal.project' persistently.",
+  commandNotes        = Just $ \pname ->
+        "Examples:\n"
+     ++ "  " ++ pname ++ " v2-reconfigure --with-compiler ghc-7.10.3\n"
+     ++ "    Adjust the project configuration to use the given compiler\n"
+     ++ "    program and check the resulting configuration works.\n"
+     ++ "  " ++ pname ++ " v2-reconfigure\n"
+     ++ "    Reset the local configuration to empty and check the overall\n"
+     ++ "    project configuration works.\n"
+
+  , commandDefaultFlags = defaultNixStyleFlags ()
+  , commandOptions      = filter (\o -> optionName o /= "ignore-project")
+                        . nixStyleOptions (const [])
+  }
+
+reconfigureAction :: NixStyleFlags () -> [String] -> GlobalFlags -> IO ()
+reconfigureAction flags extraArgs globalFlags = do
+  reconf <- reconfigureAction' flags extraArgs globalFlags
+  let go (baseCtx, projConfig) = 
+        writeProjectLocalExtraConfig (distDirLayout baseCtx) projConfig
+
+  either go go reconf
+
+reconfigureAction' 
+    :: NixStyleFlags () 
+    -> [String] 
+    -> GlobalFlags 
+    -> IO (Either (ProjectBaseContext, ProjectConfig) (ProjectBaseContext, ProjectConfig))
+reconfigureAction' flags@NixStyleFlags {..} extraArgs globalFlags = do
+  baseCtx <- establishProjectBaseContext v cliConfig OtherCommand
+
+  let localFile = distProjectFile (distDirLayout baseCtx) "local"
+
+  exists <- doesFileExist localFile
+  if not exists
+    then do
+      notice v $
+        quote (takeFileName localFile)
+        <> " doesn't exist, calling configure instead."
+
+      -- @configureAction'@ is called, as opposed to @configureAction@, to avoid
+      -- calling on the writing function for configure, thusly maintaining the 
+      -- separation on the behaviour that modify files
+      Left <$> configureAction' flags extraArgs globalFlags
+  
+    else do
+      conf <- runRebuild (distProjectRootDirectory . distDirLayout $ baseCtx) $
+        readProjectLocalExtraConfig v (distDirLayout baseCtx)
+
+      buildCtx <- runProjectPreBuildPhase v baseCtx $ \plan -> 
+        return (plan, Map.empty)
+
+      printPlan v 
+        (baseCtx {buildSettings = (buildSettings baseCtx) {buildSettingDryRun = True}})
+        buildCtx
+
+      return $ Right (baseCtx, configurationMerge conf cliConfig)
+      
+  where
+    v = fromFlagOrDefault normal (configVerbosity configFlags)
+    cliConfig = commandLineFlagsToProjectConfig globalFlags flags mempty
+    quote s = "'" <> s <> "'"
+
+configurationMerge
+    :: ProjectConfig -- The current configuration
+    -> ProjectConfig -- The configuration to merge
+    -> ProjectConfig
+configurationMerge config cliConfig =
+    config <> cliConfig
```

Comment:
This doesn't meet my Fairbarn threshold. It makes more sense just to use `<>` and not introduce a new name.

Reply:
I thought about that, but decided to go this route to make it easier to modify it afterwards, if a more complex kind of configuration merging takes place. I have no strong feelings on this, this function can be dissolved into the other one, if desired.

Reply:
I don't think we'll have a more complicated merge than this one :smile: